File: src/TensorOptions.cpp
#include <iostream>

#define LANTERN_BUILD

#include "lantern/lantern.h"

#include <torch/torch.h>

#include "utils.hpp"

void* lantern_TensorOptions () {
  return (void *) new LanternObject<torch::TensorOptions>(torch::TensorOptions());
}

void* lantern_TensorOptions_dtype (void* self, void* dtype) {
  auto out = reinterpret_cast<torch::TensorOptions * >(self)->dtype(*reinterpret_cast<torch::Dtype *>(dtype));
  return (void *) new LanternObject<torch::TensorOptions>(out);
}

void* lantern_TensorOptions_layout (void* self, void* layout) {
  auto out = reinterpret_cast<torch::TensorOptions * >(self)->layout(*reinterpret_cast<torch::Layout *>(layout));
  return (void *) new LanternObject<torch::TensorOptions>(out);
}

void* lantern_TensorOptions_device (void* self, void* device) {
  auto out = reinterpret_cast<torch::TensorOptions * >(self)->device(*reinterpret_cast<torch::Device *>(device));
  return (void *) new LanternObject<torch::TensorOptions>(out);
}

void* lantern_TensorOptions_requires_grad (void* self, bool requires_grad) {
  auto out = reinterpret_cast<torch::TensorOptions * >(self)->requires_grad(requires_grad);
  return (void *) new LanternObject<torch::TensorOptions>(out);
}

void* lantern_TensorOptions_pinned_memory (void* self, bool pinned_memory) {
  auto out = reinterpret_cast<torch::TensorOptions * >(self)->pinned_memory(pinned_memory);
  return (void *) new LanternObject<torch::TensorOptions>(out);
}

File: R/tensor_options.R


torch_tensor_options <- function(dtype = NULL, layout = NULL, device = NULL,
                                 requires_grad = NULL, pinned_memory = NULL) {
  options <- list(
    dtype = if (is.character(dtype)) dtype else dtype$ptr,
    layout = layout$ptr,
    device = if (is.character(device)) torch_device(device)$ptr else device$ptr,
    requires_grad = requires_grad,
    pinned_memory = pinned_memory
  )
  options <- Filter(Negate(is.null), options)
  options
}

is_torch_tensor_options <- function(x) {
  inherits(x, "torch_tensor_options")
}

as_torch_tensor_options <- function(l) {
  do.call(torch_tensor_options, l)
}
